internal/collectors/container_runtime.go

```diff
@@ -0,0 +1,76 @@
+package collectors
+
+import (
+	"os"
+	"regexp"
+
+	"github.com/SUSE/connect-ng/internal/util"
+	"github.com/genuinetools/bpfd/proc"
+)
+
+type ContainerRuntime struct{}
+
+func (cr ContainerRuntime) run(arch string) (Result, error) {
+	runtime := proc.GetContainerRuntime(0, 0)
+
+	// The `proc` package from `bpfd` is known to miss on detecting Docker
+	// images that were built outside of the Docker ecosystem (e.g. SUSE's
+	// kiwi). For that reason the people from
+	// https://github.com/GoogleContainerTools/kaniko forked this package and
+	// added further detection. However, we cannot simply pull this package from
+	// kaniko because they explicitely require Go 1.22, and that is something
+	// that we might not be able to afford right now. Hence, let's add the
+	// missing bits here, which aren't too large anyways.
+	if runtime == proc.RuntimeNotFound {
+		// Some files might be telling of the container runtime being used.
+		runtime = detectContainerFiles()
+		if runtime != proc.RuntimeNotFound {
+			return Result{"container-runtime": runtime}, nil
+		}
+
+		// Docker was not detected at this point. An overlay mount on "/" may
+		// indicate we're under containerd or other runtime.
+		contents := readFileString("/proc/mounts")
+		if m, _ := regexp.MatchString("^[^ ]+ / overlay", contents); m {
+			return Result{"container-runtime": proc.RuntimeKubernetes}, nil
+		}
+
+		// Everything has been exhausted, let's return a no result.
+		return NoResult, nil
+	}
+
+	return Result{"container-runtime": runtime}, nil
+}
+
+// Taken from: https://github.com/GoogleContainerTools/kaniko/blob/ad1896a680f9bf26a4388ba0e702e7aac91fddc2/pkg/util/proc/proc.go#L153-L175
+// Related implementation: https://github.com/systemd/systemd/blob/6604fb0207ee10e8dc05d67f6fe45de0b193b5c4/src/basic/virt.c#L523-L549
+func detectContainerFiles() proc.ContainerRuntime {
+	files := []struct {
+		runtime  proc.ContainerRuntime
+		location string
+	}{
+		// https://github.com/containers/podman/issues/6192
+		// https://github.com/containers/podman/issues/3586#issuecomment-661918679
+		{proc.RuntimePodman, "/run/.containerenv"},
+		// https://github.com/moby/moby/issues/18355
+		{proc.RuntimeDocker, "/.dockerenv"},
+		// Detect the presence of a serviceaccount secret mounted in the default location
+		{proc.RuntimeKubernetes, "/var/run/secrets/kubernetes.io/serviceaccount"},
+	}
+
+	for i := range files {
+		if util.FileExists(files[i].location) {
+			return files[i].runtime
+		}
+	}
+	return proc.RuntimeNotFound
+}
+
+// Adapted from: https://github.com/genuinetools/bpfd/blob/c12d8cd9aac8674c52b83e60e6e762fa59f4e96b/proc/proc.go.
+func readFileString(file string) string {
```

Comment:
Can we have this in a helper file? So we can also mock it in case we want to test something